examples/fastapi-embeddings-rag/app/main.py

```diff
@@ -0,0 +1,379 @@
+#
+# ![Graphic displaying the steps of indexing data and the retrieval and generation process](https://runhouse-tutorials.s3.amazonaws.com/indexing-retrieval-generation.png)
+#
+# Note: Some of the steps in this example could also be accomplished with platforms like OpenAI and
+# tools such as LangChain, but we break out the components explicitly to fully illustrate each step and make the
+# example easily adaptible to other use cases. Swap out components as you see fit!
+#
+# ### What does Runhouse enable?
+# Runhouse allows you to turn complex operations such as preprocessing and inference into independent services.
+# Servicifying with Runhouse enables:
+# - **Decoupling**: AI/ML or compute-heavy heavy tasks can be separated from your main application.
+#   This keeps the FastAPI app light and allows each service to scale independently.
+# - **Multi-cloud**: Host Runhouse modules on remote machines from any cloud provider (even on the same GPU).
+#   Take advantage of unused cloud credits and avoid platform dependence.
+# - **Sharing**: Running on GPUs can be expensive, especially if they aren't fully utilized. By sharing services within
+#   your organization, you can substantially cut costs.
+#
+# ### Why FastAPI?
+# We chose FastAPI as our platform because of its popularity and simplicity. However, we could
+# easily use any other *Python-based* platform with Runhouse. Streamlit, Flask, `<your_favorite>`, we got you!
+#
+# ## Setup credentials and dependencies
+#
+# To ensure that Runhouse is able to manage deploying services to your cloud provider (AWS in this case)
+# you may need to follow initial setup steps. Please visit the AWS section of
+# our [Installation Guide](https://www.run.house/docs/installation)
+#
+# Additionally, we'll be downloading the Llama 3 model from Hugging Face, so we need to set up our Hugging Face token:
+# ```shell
+# $ export HF_TOKEN=<your huggingface token>
+# ```
+#
+# Make sure to sign the waiver on the [Hugging Face model page](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct)
+# so that you can access it.
+#
+# ## FastAPI RAG App Setup
+# First, we'll import necessary packages and initialize variables used in the application. The `URLEmbedder` and
+# `LlamaModel` classes that will be sent to Runhouse are available in the `app/modules` folder in this source code.
+from contextlib import asynccontextmanager
+from typing import Dict, List
+
+import lancedb
+
+import runhouse as rh
+
+from fastapi import Body, FastAPI, HTTPException
+
+from app.modules.embedding import Item, URLEmbedder
+from app.modules.llm import LlamaModel
+
+EMBEDDER, TABLE, LLM = None, None, None
+
+DEBUG = True  # In DEBUG mode we will always override Runhouse modules
+
+# Configuration options for our remote cluster
+CLUSTER_NAME = "rh-xa10g"  # Allows the cluster to be reused
+INSTANCE_TYPE = "A10G:1"  # A10G GPU to handle LLM iinference
+CLOUD_PROVIDER = "aws"  # Alternatively "gcp", "azure", or "cheapest"
```

Comment:
thoughts on a note to install runhouse with the relevant subdependency? (we're specifying `runhouse[aws]` in the reqs)